glide-core/redis-rs/redis/src/commands/cluster_scan.rs

```diff
@@ -323,51 +367,73 @@ impl ScanState {
-            }
-            Err(err) => Err(err),
+
+        if *new_address == NO_ADDRESSES_FOUND {
+            return Ok(ScanState::create_finished_state());
         }
+        let new_epoch = connection
+            .get_address_epoch(&new_address)
+            .await
+            .unwrap_or(0);
+        let new_cursor = 0;
+        Ok(ScanState::new(
+            new_cursor,
+            scanned_slots_map,
+            new_address,
+            new_epoch,
+            ScanStateStage::InProgress,
+        ))
     }
 }
 
 // Implement the [`ClusterInScan`] trait for [`InnerCore`] of async cluster connection.
-#[async_trait]
 impl<C> ClusterInScan for Core<C>
 where
     C: ConnectionLike + Connect + Clone + Send + Sync + 'static,
 {
-    async fn get_address_by_slot(&self, slot: u16) -> RedisResult<Arc<String>> {
-        let address = self
-            .get_address_from_slot(slot, SlotAddr::ReplicaRequired)
-            .await;
-        match address {
-            Some(addr) => Ok(addr),
-            None => {
-                if self.are_all_slots_covered().await {
-                    Err(RedisError::from((
+    async fn get_address_by_slot(
+        &self,
+        mut slot: u16,
+        scanned_slots_map: &mut SlotsBitsArray,
+        allow_non_covered_slots: bool,
+    ) -> RedisResult<Arc<String>> {
+        loop {
+            let address = self
+                .get_address_from_slot(slot, SlotAddr::ReplicaRequired)
+                .await;
+
+            match address {
+                Some(addr) => return Ok(addr),
+                None => {
+                    if allow_non_covered_slots {
+                        // Mark the current slot as scanned before moving to next
+                        let slot_index = slot as usize / BITS_PER_U64;
+                        let slot_bit = slot as usize % BITS_PER_U64;
+                        scanned_slots_map[slot_index] |= 1 << slot_bit;
+
+                        slot += 1;
+                        if slot == NUM_OF_SLOTS as u16 {
+                            return Ok(NO_ADDRESSES_FOUND.to_string().into());
+                        }
+                        continue;
+                    }
+
+                    if self.are_all_slots_covered().await {
+                        return
+                        Err::<Arc<String>, RedisError>(RedisError::from((
                         ErrorKind::IoError,
                         "Failed to get connection to the node cover the slot, please check the cluster configuration ",
-                    )))
-                } else {
-                    Err(RedisError::from((
+                    )));
+                    }
+                    return
+                        Err::<Arc<String>, RedisError>(RedisError::from((
                         ErrorKind::NotAllSlotsCovered,
                         "All slots are not covered by the cluster, please check the cluster configuration ",
-                    )))
+                    )));
```

Comment:
I don't understand why are you checking are_all_slots_covered. If we received None for slot-for-address, it means that we don't have a connection to this node at the moment, e.g. it isn't covered. what do you mean by " check the cluster configuration"? why can't we simply return  ErrorKind::NotAllSlotsCovered,?

e.g.
```
enum NextNodeResult {
     Address(Arc<String>),
     AllSlotsDone,
}

fn get_next_node(slot, ...) -> RedisResult<NextNodeResult>
     while slot < NUM_OF_SLOTS {
           match self.get_address_from_slot(slot, SlotAddr::ReplicaRequired).await {
                Some(addr) => return Ok(NextNodeResult::Address(addr))),
                None => {
                       if allow_non_covered_slots {
                                     mark_slot_as_done(slot);
                                     slot++;
                       } else {
                                      return Err(NotAllSlotsCovered)
                       }
                }
      }
    NextNodeResult::AllSlotsDone
}

Reply:
The message was decided with you and Asaf actually.
I will change it, but i want something a bit more specific.